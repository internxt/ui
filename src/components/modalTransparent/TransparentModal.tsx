import { ReactNode, useEffect, useRef, useState } from 'react';

export interface TransparentModalProps {
  isOpen: boolean;
  onClose: () => void;
  children: ReactNode;
  className?: string;
  disableBackdrop?: boolean;
}

const TransparentModal = ({ isOpen, onClose, children, className, disableBackdrop = false }: TransparentModalProps) => {
  const modalRef = useRef<HTMLDivElement>(null);
  const [showContent, setShowContent] = useState(isOpen);
  const [transitionOpacity, setTransitionOpacity] = useState('opacity-0');
  const [transitionScale, setTransitionScale] = useState('scale-95');

  useEffect(() => {
    if (isOpen) {
      setShowContent(true);
      const timeout = setTimeout(() => {
        setTransitionOpacity('opacity-100');
        setTransitionScale('scale-100');
      }, 10);
      return () => clearTimeout(timeout);
    } else {
      setTransitionOpacity('opacity-0');
      setTransitionScale('scale-95');
      const timeout = setTimeout(() => {
        setShowContent(false);
      }, 150);
      return () => clearTimeout(timeout);
    }
  }, [isOpen]);

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (modalRef.current && !modalRef.current.contains(event.target as Node)) {
        onClose();
      }
    };

    if (isOpen && !disableBackdrop) {
      document.addEventListener('mousedown', handleClickOutside);
    }

    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [isOpen, onClose, disableBackdrop]);

  if (!showContent) return null;

  return (
    <div
      className={`
      fixed 
      inset-0 
      z-50 
      flex 
      items-center 
      justify-center
      ${disableBackdrop ? 'pointer-events-none' : ''}
    `}
    >
      {/* Backdrop */}
      {!disableBackdrop && (
        <div
          className={`
            fixed 
            inset-0 
            bg-black/50 
            backdrop-blur-sm 
            transition-opacity 
            duration-200
            ${transitionOpacity}
          `}
        />
      )}

      {/* Modal */}
      <div
        ref={modalRef}
        className={`
          relative
          flex
          bg-black/15
          border
          border-white/15
          rounded-[20px]
          backdrop-blur
          shadow-lg
          transition-all
          duration-200
          pointer-events-auto
          ${transitionOpacity}
          ${transitionScale}
          ${className}
        `}
      >
        {children}
      </div>
    </div>
  );
};

export default TransparentModal;
