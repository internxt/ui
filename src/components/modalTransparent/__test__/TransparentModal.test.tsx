import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import { afterAll, afterEach, beforeAll, describe, expect, it, vi } from 'vitest';
import TransparentModal from '../TransparentModal';

describe('TransparentModal Component', () => {
  const onCloseMock = vi.fn();
  const renderTransparentModal = (props = {}) => {
    return render(
      <TransparentModal isOpen={true} onClose={onCloseMock} {...props}>
        <div className="text-center">
          <h2 className="text-lg font-semibold">Title</h2>
          <p className="mt-2">Modal Content</p>
          <button className="mt-4 px-4 py-2 bg-blue-600 text-white rounded hover:bg-blue-700" onClick={onCloseMock}>
            Close
          </button>
        </div>
      </TransparentModal>,
    );
  };

  beforeAll(() => {
    (global as any).IntersectionObserver = class {
      observe() {}
      unobserve() {}
      disconnect() {}
    };
  });

  afterEach(() => {
    vi.clearAllMocks();
    vi.restoreAllMocks();
  });

  afterAll(() => {
    delete (global as any).IntersectionObserver;
  });

  it('should match snapshot when isOpen is true', () => {
    const { container } = renderTransparentModal();
    expect(container).toMatchSnapshot();
  });

  it('should render the modal when isOpen is true', () => {
    renderTransparentModal();
    expect(screen.getByText('Modal Content')).toBeInTheDocument();
  });

  it('should not render the modal when isOpen is false', () => {
    render(
      <TransparentModal isOpen={false} onClose={onCloseMock}>
        <div>Modal Content</div>
      </TransparentModal>,
    );
    expect(screen.queryByText('Modal Content')).not.toBeInTheDocument();
  });

  it('should call onClose when clicking outside if disableBackdrop is false', () => {
    renderTransparentModal();
    fireEvent.mouseDown(document.body);
    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  it('should not call onClose when clicking outside if disableBackdrop is true', () => {
    renderTransparentModal({ disableBackdrop: true });
    fireEvent.mouseDown(document.body);
    expect(onCloseMock).not.toHaveBeenCalled();
  });

  it('should apply custom class names', () => {
    render(
      <TransparentModal isOpen={true} onClose={onCloseMock} className="custom-class" disableBackdrop>
        Modal Content
      </TransparentModal>,
    );
    const modalElement = screen.getByText('Modal Content');
    expect(modalElement).toHaveClass('custom-class');
  });

  it('should show backdrop when disableBackdrop is false', () => {
    renderTransparentModal();
    const backdrop = document.querySelector('.bg-black\\/50');
    expect(backdrop).toBeInTheDocument();
  });

  it('should not show backdrop when disableBackdrop is true', () => {
    renderTransparentModal({ disableBackdrop: true });
    const backdrop = document.querySelector('.bg-black\\/50');
    expect(backdrop).not.toBeInTheDocument();
  });

  it('should update transition classes after timeout when isOpen is true', async () => {
    renderTransparentModal();
    const modalContainer = document.querySelector('.relative.flex.bg-black\\/15');

    expect(modalContainer).toHaveClass('opacity-0');
    expect(modalContainer).toHaveClass('scale-95');

    await waitFor(() => {
      expect(modalContainer).toHaveClass('opacity-100');
      expect(modalContainer).toHaveClass('scale-100');
    });
  });

  it('should reset transition classes when isOpen changes to false', async () => {
    const { rerender } = renderTransparentModal();
    const modalContainer = document.querySelector('.relative.flex.bg-black\\/15');

    await waitFor(() => {
      expect(modalContainer).toHaveClass('opacity-100');
      expect(modalContainer).toHaveClass('scale-100');
    });

    rerender(
      <TransparentModal isOpen={false} onClose={onCloseMock}>
        <div>Modal Content</div>
      </TransparentModal>,
    );

    expect(modalContainer).toHaveClass('opacity-0');
    expect(modalContainer).toHaveClass('scale-95');

    await waitFor(() => {
      expect(screen.queryByText('Modal Content')).not.toBeInTheDocument();
    });
  });

  it('should add pointer-events-none class when disableBackdrop is true', () => {
    renderTransparentModal({ disableBackdrop: true });
    const container = document.querySelector('.fixed.inset-0.z-50');
    expect(container).toHaveClass('pointer-events-none');
  });
});
