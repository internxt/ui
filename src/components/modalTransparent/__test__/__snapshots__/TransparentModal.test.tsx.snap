// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TransparentModal Component > should match snapshot when isOpen is true 1`] = `
<div>
  <div
    class="
      fixed 
      inset-0 
      z-50 
      flex 
      items-center 
      justify-center
      
    "
  >
    <div
      class="
            fixed 
            inset-0 
            bg-black/50 
            backdrop-blur-sm 
            transition-opacity 
            duration-200
            opacity-0
          "
    />
    <div
      class="
          relative
          flex
          bg-black/15
          border
          border-white/15
          rounded-[20px]
          backdrop-blur
          shadow-lg
          transition-all
          duration-200
          pointer-events-auto
          opacity-0
          scale-95
          undefined
        "
    >
      <div
        class="text-center"
      >
        <h2
          class="text-lg font-semibold"
        >
          Title
        </h2>
        <p
          class="mt-2"
        >
          Modal Content
        </p>
        <button
          class="mt-4 px-4 py-2 bg-blue-600 text-white rounded hover:bg-blue-700"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;
